File: backend/app/schemas/portfolio.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any


# ---------------------------------------------------------------------------
# Modelos existentes (compatibilidad con frontend actual)
# ---------------------------------------------------------------------------

class FundBase(BaseModel):
    Fondo: str
    TIPO: str
    Porcentaje: float = 0.0
    ISIN: Optional[str] = None

    # Datos extendidos que rellenan los analyzers
    NAV_Precio: Optional[str] = Field(None, alias="NAV (Precio)")
    YTD: Optional[str] = Field(None, alias="YTD (%)")
    Morningstar_Name: Optional[str] = Field(None, alias="Morningstar Name")
    Estrellas_MS: Optional[str] = Field(None, alias="Estrellas MS")
    Categoria: Optional[str] = Field(None, alias="Categoría")
    IsIndex: Optional[bool] = None
    Valor_Actual: Optional[float] = None
    Capital_Invertido: Optional[float] = None
    Participaciones: Optional[float] = None
    Fecha_Compra: Optional[str] = None
    Ganancia_Abs: Optional[float] = None
    Ganancia_Pct: Optional[float] = None
    finect_url: Optional[str] = None  # Full Finect URL with slug

    class Config:
        populate_by_name = True


class PortfolioSummary(BaseModel):
    total_rv: float
    total_rf: float
    total_cash: float
    total_alt: float
    total_indexed: Optional[float] = 0.0
    total_active: Optional[float] = 0.0
    details: Dict[str, float]


class AnalysisResponse(BaseModel):
    summary: PortfolioSummary
    funds: List[FundBase]
    recommendation: Dict[str, Dict[str, str]]
    real_evolution: Optional[Dict[str, Any]] = None
    built_at: Optional[str] = None  # ISO timestamp when this response was built


# ---------------------------------------------------------------------------
# Nuevos modelos
# ---------------------------------------------------------------------------

class PositionItem(BaseModel):
    ISIN: str
    Fondo: str
    Participaciones: float
    Precio_Compra_Medio: float = 0.0
    Capital_Invertido: float = 0.0
    Precio_Actual: Optional[float] = None
    Valor_Actual: Optional[float] = None
    Ganancia_Euros: Optional[float] = None
    Ganancia_Pct: Optional[float] = None
    finect_url: Optional[str] = None  # Full Finect URL with slug


class PositionsResponse(BaseModel):
    positions: List[PositionItem]
    total_invested: float
    total_value: float
    total_gain: float
    total_gain_pct: float


class OpenLotItem(BaseModel):
    ISIN: str
    Fondo: str
    Fecha_Compra: Optional[str] = None
    Participaciones_Iniciales: float = 0.0
    Participaciones_Restantes: float = 0.0
    Importe_Invertido: float = 0.0
    Precio_Compra_Unitario: float = 0.0


class TaxOptimizeRequest(BaseModel):
    target_amount: float = Field(..., gt=0, description="Cantidad a retirar en €")


class TaxPlanStep(BaseModel):
    ISIN: str
    Fondo: str
    Fecha_Compra: Optional[str] = None
    Participaciones_Vendidas: float
    Importe_Retirado: float
    Ganancia_Patrimonial: float
    es_etf: bool = False  # True si es ETF/ETP (no traspasable en España)


class TaxOptimizeResponse(BaseModel):
    target_amount: float
    withdrawn_amount: float
    total_capital_gain: float
    estimated_tax: float
    net_amount: float
    plan: List[TaxPlanStep]


class FundMetrics(BaseModel):
    """Métricas de riesgo/rendimiento de un fondo (fuente: Finect)."""
    sharpe_ratio: Optional[float] = None
    alpha: Optional[float] = None
    beta: Optional[float] = None
    standard_deviation: Optional[float] = None
    max_drawdown: Optional[float] = None
    tracking_error: Optional[float] = None
    information_ratio: Optional[float] = None
    r2: Optional[float] = None
    correlation: Optional[float] = None


class FundDetailResponse(BaseModel):
    isin: str
    name: str = ""
    expense_ratio: Optional[float] = None
    aum: Optional[float] = None
    inception_date: Optional[str] = None
    rating: Optional[Any] = None
    risk_score: Optional[Any] = None
    srri: Optional[int] = None
    category: Optional[str] = None
    management_company: Optional[str] = None
    metrics: Optional[FundMetrics] = None
    sectors: Dict[str, float] = {}
    countries: Dict[str, float] = {}
    asset_allocation: Dict[str, float] = {}
    market_cap: Dict[str, float] = {}
    holdings: List[Dict[str, Any]] = []
    source: str = ""
    finect_url: Optional[str] = None


class FundSearchResult(BaseModel):
    """Resultado de búsqueda de un fondo en Finect."""
    isin: str
    name: str = ""
    category: Optional[str] = None
    management_company: Optional[str] = None
    in_portfolio: bool = False
    url: Optional[str] = None  # Full Finect URL with slug
    ticker: Optional[str] = None  # Exchange ticker (e.g. NUKL, WNUC)


class TraspasoFundItem(BaseModel):
    """Análisis de un fondo respecto a la optimización por traspaso."""
    isin: str
    nombre: str = ""
    valor_actual: float = 0.0
    capital_invertido: float = 0.0
    plusvalia_latente: float = 0.0
    plusvalia_pct: float = 0.0
    impuesto_si_vendes: float = 0.0
    ahorro_traspaso: float = 0.0
    num_lotes: int = 0
    cualifica_traspaso: bool = True
    is_etf: bool = False  # True si es ETF/ETP (no traspasable en España)


# ---------------------------------------------------------------------------
# Optimización de retirada via traspaso (greedy global FIFO)
# ---------------------------------------------------------------------------

class TraspasoOptimizeRequest(BaseModel):
    """Petición de optimización de retirada vía traspaso."""
    target_amount: float = Field(..., gt=0, description="Cantidad a retirar en €")


class TraspasoLotStep(BaseModel):
    """Detalle de un lote en el plan de traspaso o reembolso."""
    ISIN: str
    Fondo: str = ""
    Fecha_Compra: Optional[str] = None
    Participaciones: float = 0.0
    # Reembolso
    Importe: Optional[float] = None
    Ganancia_Patrimonial: Optional[float] = None
    # Traspaso
    Importe_Traspasado: Optional[float] = None
    Plusvalia_Diferida: Optional[float] = None
    Destination_ISIN: Optional[str] = None
    Destination_Fondo: Optional[str] = None
    Precio_Compra_Unitario: float = 0.0
    Nota: Optional[str] = None
    es_etf: bool = False  # True si es ETF/ETP (no traspasable en España)


class EscenarioFiscal(BaseModel):
    """Resultado fiscal de un escenario (directo u optimizado)."""
    ganancia_patrimonial: float = 0.0
    impuesto: float = 0.0
    withdrawn_amount: float = 0.0
    neto_recibido: float = 0.0
    detalle: List[TraspasoLotStep] = []


class DestinationFund(BaseModel):
    """Fondo destino para los traspasos."""
    isin: str
    nombre: str = ""
    tipo: str = "new_suggestion"   # "portfolio_index" | "new_suggestion"
    is_index: bool = True
    motivo: str = ""


class LossHarvestingCandidate(BaseModel):
    """Un lote candidato a tax-loss harvesting."""
    ISIN: str
    Fondo: str = ""
    es_etf: bool = False
    Fecha_Compra: Optional[str] = None
    lot_loss: float = 0.0
    lot_value: float = 0.0
    preceding_forced_gain: float = 0.0
    preceding_forced_value: float = 0.0
    preceding_transfer_value: float = 0.0
    net_harvest_gain: float = 0.0
    additional_cash: float = 0.0
    antiaplicacion_plazo: str = ""


class LossHarvestingSuggestion(BaseModel):
    """Sugerencia bidireccional de harvesting fiscal.

    direction:
      - ``harvest_losses``: vender lotes en pérdida para compensar ganancias del plan.
      - ``harvest_gains``: vender lotes con ganancia aprovechando pérdidas del plan →
        dinero extra sin coste fiscal.
      - ``none``: sin sugerencia (ganancia neta = 0).
    """
    direction: str = "none"
    candidates: List[LossHarvestingCandidate] = []
    base_net_gain: float = 0.0
    base_tax: float = 0.0
    total_harvestable_loss: float = 0.0
    net_gain_after_harvest: float = 0.0
    tax_after_harvest: float = 0.0
    tax_savings: float = 0.0
    additional_cash: float = 0.0


class TraspasoOptimizeResponse(BaseModel):
    """Respuesta completa de la optimización global de retirada vía traspaso."""
    target_amount: float
    total_portfolio_value: float = 0.0
    escenario_directo: EscenarioFiscal
    escenario_optimizado: EscenarioFiscal
    ahorro_fiscal: float = 0.0
    ahorro_fiscal_pct: float = 0.0
    plan_traspasos: List[TraspasoLotStep] = []
    plan_reembolso: List[TraspasoLotStep] = []
    importe_traspasado: float = 0.0
    plusvalia_diferida: float = 0.0
    fondos_afectados: List[str] = []
    destination_fund: Optional[DestinationFund] = None
    destination_alternatives: List[Dict[str, Any]] = []
    portfolio_after: List[Dict[str, Any]] = []
    non_traspasable_isins: List[str] = []  # ETFs/ETPs en cartera (no traspasables)
    loss_harvesting: Optional[LossHarvestingSuggestion] = None
    notas: str = ""


class SimulationRequest(BaseModel):
    """Petición de simulación: añadir X€ a un fondo."""
    isin: str
    amount: float = Field(..., gt=0, description="Cantidad a añadir en €")


class SimulatedFundDetail(BaseModel):
    """Detalle de un fondo en la simulación."""
    isin: str
    name: str = ""
    current_weight: float = 0.0
    simulated_weight: float = 0.0
    amount: float = 0.0
    metrics: Optional[FundMetrics] = None


class SimulationResponse(BaseModel):
    """Resultado de la simulación de añadir dinero a un fondo."""
    added_isin: str
    added_name: str = ""
    added_amount: float
    current_total: float
    simulated_total: float
    current_portfolio_metrics: Dict[str, Optional[float]] = {}
    simulated_portfolio_metrics: Dict[str, Optional[float]] = {}
    funds: List[SimulatedFundDetail] = []
    # Historical series for chart [{date, price}] — base 100
    history_current: List[Dict[str, Any]] = []
    history_fund: List[Dict[str, Any]] = []
    history_simulated: List[Dict[str, Any]] = []
    # Period returns [{label, current, fund, simulated}] — % total or CAGR
    period_returns: List[Dict[str, Any]] = []


class RebalanceFundDetail(BaseModel):
    """Movimiento necesario en el rebalanceo para un fondo."""
    isin: str
    name: str = ""
    current_weight: float = 0.0
    target_weight: float = 0.0
    delta_eur: float = 0.0


class RebalanceRequest(BaseModel):
    """Petición de rebalanceo: pesos objetivo por ISIN (fracción, suma = 1)."""
    weights: Dict[str, float] = Field(..., description="ISIN → fracción objetivo (0-1)")


class RebalanceResponse(BaseModel):
    """Resultado del rebalanceo simulado."""
    total_value: float
    current_portfolio_metrics: Dict[str, Optional[float]] = {}
    simulated_portfolio_metrics: Dict[str, Optional[float]] = {}
    funds: List[RebalanceFundDetail] = []
    history_current: List[Dict[str, Any]] = []
    history_simulated: List[Dict[str, Any]] = []
    period_returns: List[Dict[str, Any]] = []

File: frontend/src/types/portfolio.ts
/* ─── TypeScript types mirroring backend schemas ─────────────────────────── */

/** Fund basic information (matches backend FundBase) */
export interface Fund {
  Fondo: string;
  TIPO: string;
  Porcentaje: number;
  ISIN?: string;
  "NAV (Precio)"?: string;
  "YTD (%)"?: string;
  "Morningstar Name"?: string;
  "Estrellas MS"?: string;
  Categoría?: string;
  IsIndex?: boolean;
  Valor_Actual?: number;
  Capital_Invertido?: number;
  Ganancia_Abs?: number;
  Ganancia_Pct?: number;
  finect_url?: string;
}

/** Portfolio summary breakdown */
export interface PortfolioSummary {
  total_rv: number;
  total_rf: number;
  total_cash: number;
  total_alt: number;
  total_indexed?: number;
  total_active?: number;
  details: Record<string, number>;
}

/** Main API response for /api/portfolio/summary */
export interface AnalysisResponse {
  summary: PortfolioSummary;
  funds: Fund[];
  recommendation: Record<string, Record<string, string>>;
  real_evolution?: RealEvolution;
  built_at?: string;
}

/** Real evolution data from /api/portfolio/real-evolution */
export interface RealEvolution {
  series: EvolutionPoint[];
  monthly: MonthlyData[];
  per_fund?: Record<string, PerFundPoint[]>;
  funds?: Record<string, Array<{ date: string; value: number }>>;
  invested_per_fund?: Record<string, Array<{ date: string; invested: number }>>;
  monthly_per_fund?: Record<
    string,
    Array<{
      date: string;
      value: number;
      invested: number;
      gain: number;
      gain_pct: number;
    }>
  >;
}

export interface EvolutionPoint {
  date: string;
  value: number;
  invested: number;
  gain?: number;
  gain_pct?: number;
}

export interface MonthlyData {
  month: string;
  date?: string;
  label?: string;
  value: number;
  invested: number;
  gain: number;
  gain_pct: number;
  contributions: number;
  mom?: number;
}

export interface PerFundPoint {
  date: string;
  value: number;
  invested?: number;
}

/** Position item from /api/portfolio/positions */
export interface PositionItem {
  ISIN: string;
  Fondo: string;
  Participaciones: number;
  Precio_Compra_Medio: number;
  Capital_Invertido: number;
  Precio_Actual?: number;
  Valor_Actual?: number;
  Ganancia_Euros?: number;
  Ganancia_Pct?: number;
  finect_url?: string;
}

/** Fund detail response from /api/portfolio/fund/{isin}/details */
export interface FundDetail {
  isin: string;
  name: string;
  expense_ratio?: number;
  aum?: number;
  inception_date?: string;
  rating?: unknown;
  risk_score?: unknown;
  srri?: number;
  category?: string;
  management_company?: string;
  metrics?: FundMetrics;
  sectors: Record<string, number>;
  countries: Record<string, number>;
  asset_allocation: Record<string, number>;
  market_cap: Record<string, number>;
  holdings: Array<Record<string, unknown>>;
  source: string;
  finect_url?: string;
}

/** Fund risk/return metrics */
export interface FundMetrics {
  sharpe_ratio?: number;
  alpha?: number;
  beta?: number;
  standard_deviation?: number;
  max_drawdown?: number;
  tracking_error?: number;
  information_ratio?: number;
  r2?: number;
  correlation?: number;
}

/** Fund search result */
export interface FundSearchResult {
  isin: string;
  name: string;
  category?: string;
  management_company?: string;
  in_portfolio: boolean;
  url?: string;
  ticker?: string;
}

/** Orders summary response */
export interface OrdersSummaryResponse {
  monthly: Record<string, number>;
  yearly: Record<string, number>;
}

/** Annual returns data */
export interface AnnualReturnsData {
  annual: Record<string, Record<string, number>>;
  monthly?: Record<string, Record<string, Record<string, number>>>;
}

/** History batch response */
export interface HistoryBatchResponse {
  series: Record<string, Array<{ date: string; price: number }>>;
  base_date?: string;
}

/** Correlation matrix */
export interface CorrelationData {
  matrix: Record<string, Record<string, number>>;
  names: Record<string, string>;
}

/** Last update info */
export interface LastUpdateInfo {
  last_update: string;
  last_date?: string;
  fund_updates: Record<string, string>;
}

/** Benchmark data (MSCI World) */
export interface BenchmarkData {
  sectors?: Record<string, number>;
  regions?: Record<string, number>;
}

/** Portfolio Holdings response */
export interface PortfolioHoldingsResponse {
  holdings: Array<{
    name: string;
    weight: number;
  }>;
  total_funds: number;
  funds_with_holdings: number;
  coverage_pct: number;
}

/** Details response — keyed by fund name */
export interface FundDetailsMap {
  [fundName: string]: {
    isin?: string;
    percentage: number;
    finect_url?: string;
    sector?: Record<string, number>;
    region?: Record<string, number>;
  };
}

/* ─── Simulator types ────────────────────────────────────────────────────── */

/** POST /api/portfolio/simulate response */
export interface SimulationResult {
  current_total: number;
  contribution: number;
  updated_total: number;
  history: Record<string, Array<{ date: string; price: number }>>;
  correlation?: {
    labels: string[];
    matrix: Record<string, Record<string, number | null>>;
  };
  metrics?: Record<string, FundMetrics>;
  weight_changes?: Array<{
    fund: string;
    before: number;
    after: number;
    diff: number;
  }>;
}

/** POST /api/portfolio/rebalance request */
export interface RebalanceRequest {
  transfers: Array<{
    from_isin: string;
    to_isin: string;
    amount: number;
  }>;
  standalone_adds?: Array<{
    isin: string;
    name: string;
    amount: number;
  }>;
}

/** POST /api/portfolio/rebalance response */
export interface RebalanceResult {
  history?: Record<string, Array<{ date: string; price: number }>>;
  positions_before?: Array<{
    fund: string;
    isin: string;
    balance: number;
    weight: number;
  }>;
  positions_after?: Array<{
    fund: string;
    isin: string;
    balance: number;
    weight: number;
  }>;
  metrics_comparison?: Record<
    string,
    { current: Record<string, number>; rebalanced: Record<string, number> }
  >;
}

/* ─── Withdrawals types ──────────────────────────────────────────────────── */

/** GET /api/portfolio/traspaso-analysis response */
export interface TraspasoAnalysis {
  funds: Array<{
    fund: string;
    isin: string;
    current_value: number;
    latent_gain: number;
    tax_if_sold: number;
    transfer_savings: number;
    qualification: string;
    is_etf: boolean;
  }>;
  total_deferrable_gain: number;
  total_tax_savings: number;
}

/** POST /api/portfolio/traspaso-optimize response */
export interface TraspasoOptimizeResult {
  target_amount: number;
  total_portfolio_value: number;
  /** Escenario fiscal vendiedo directamente FIFO */
  escenario_directo: {
    ganancia_patrimonial: number;
    impuesto: number;
    withdrawn_amount?: number;
    neto_recibido: number;
    detalle: Array<Record<string, unknown>>;
  };
  /** Escenario fiscal optimizado mediante traspasos */
  escenario_optimizado: {
    ganancia_patrimonial: number;
    impuesto: number;
    withdrawn_amount?: number;
    neto_recibido: number;
    detalle: Array<Record<string, unknown>>;
  };
  ahorro_fiscal: number;
  ahorro_fiscal_pct: number;
  plan_traspasos: Array<Record<string, unknown>>;
  plan_reembolso: Array<Record<string, unknown>>;
  importe_traspasado: number;
  plusvalia_diferida: number;
  fondos_afectados: string[];
  destination_fund?: { isin: string; nombre: string; tipo: string; motivo: string } | null;
  non_traspasable_isins?: string[];  // ETFs/ETPs en cartera (no traspasables)
  /** Tax-loss / gain harvesting suggestions */
  loss_harvesting?: {
    /** "harvest_losses" = sell loss lots to offset plan gains;
     *  "harvest_gains" = sell gain lots covered by plan losses → tax-free cash;
     *  "none" = nothing to suggest */
    direction: "harvest_losses" | "harvest_gains" | "none";
    candidates: Array<{
      ISIN: string;
      Fondo: string;
      es_etf: boolean;
      Fecha_Compra?: string | null;
      lot_loss: number;
      lot_value: number;
      preceding_forced_gain: number;
      preceding_forced_value: number;
      preceding_transfer_value: number;
      net_harvest_gain: number;
      additional_cash: number;
      antiaplicacion_plazo: string;
    }>;
    base_net_gain: number;
    base_tax: number;
    total_harvestable_loss: number;
    net_gain_after_harvest: number;
    tax_after_harvest: number;
    tax_savings: number;
    additional_cash: number;
  } | null;
  notas?: string;
  // Legacy aliases (keep for backward compat)
  withdrawal_amount?: number;
  direct_fifo?: { total_gain: number; tax: number; net: number };
  optimized?: { total_gain: number; tax: number; net: number; savings: number };
}

/** POST /api/portfolio/tax-optimize response */
export interface TaxOptimizeResult {
  withdraw_amount: number;
  total_gain: number;
  estimated_tax: number;
  net_after_tax: number;
  optimal_plan: Array<{
    fund: string;
    isin?: string;
    purchase_date: string;
    shares: number;
    amount: number;
    gain: number;
    is_etf?: boolean;
  }>;
  tax_brackets?: Array<{
    bracket: string;
    rate: number;
    amount: number;
    tax: number;
  }>;
}

/* ─── Opportunities types ────────────────────────────────────────────────── */

/** GET /api/portfolio/timing-presets response */
export interface TimingPresetsResponse {
  presets: Record<
    string,
    { label: string; description: string; weights: Record<string, number> }
  >;
  default_weights: Record<string, number>;
}

/** Opportunity entry from GET /api/portfolio/opportunities */
export interface OpportunityEntry {
  isin: string;
  name: string;
  valor_actual?: number;
  ganancia_pct?: number;
  fund_type?: string;
  timing_score: number;
  trend_score: number;
  pullback_score: number;
  divergence_score: number;
  rsi_score: number;
  vol_regime_score: number;
  short_term_score: number;
  weights_used?: Record<string, number>;
  z_trend?: number;
  trend_deviation_pct?: number;
  pullback_3m_pct?: number;
  momentum_1m?: number;
  momentum_3m?: number;
  momentum_6m?: number;
  momentum_3d?: number;
  momentum_1w?: number;
  momentum_2w?: number;
  pullback_1w_pct?: number;
  pullback_2w_pct?: number;
  rsi_14?: number;
  vol_regime_ratio?: number;
  current_price?: number;
  ath?: number;
  drawdown_ath_pct?: number;
  sma200?: number;
  sma200_dist_pct?: number;
  sharpe?: number;
  sortino?: number;
  cagr_pct?: number;
  ret_1y?: number;
  ret_3y?: number;
  ret_5y?: number;
  rating?: number;
  ter_pct?: number;
  volatility_pct?: number;
  max_drawdown_pct?: number;
  consistency?: number;
  calmar?: number;
  level: string;
  description: string;
}

/** Chart data for a single opportunity fund */
export interface OpportunityChartData {
  isin: string;
  name: string;
  fund_type?: string;
  chart: {
    price_series: Array<{ date: string; price: number }>;
    regression?: Array<{ date: string; value: number }>;
    band_1_upper?: Array<{ date: string; value: number }>;
    band_1_lower?: Array<{ date: string; value: number }>;
    band_2_upper?: Array<{ date: string; value: number }>;
    band_2_lower?: Array<{ date: string; value: number }>;
    sma200?: Array<{ date: string; value: number }>;
    pullback_levels?: Record<string, number>;
    rsi_series?: Array<{ date: string; value: number }>;
    crossovers?: Array<{ date: string; type: string }>;
    chart_start?: string;
    chart_end?: string;
    std_residual?: number;
  };
  signals?: Record<string, unknown>;
  level?: string;
  description?: string;
}

/** Fund enrichment result */
export interface EnrichedFund {
  isin: string;
  name: string;
  category?: string;
  expense_ratio?: number;
  aum?: number;
  rating?: number;
  srri?: number;
  management_company?: string;
  fund_type?: string;
  returns?: Record<string, number>;
  metrics?: Record<string, number | null>;
  signals?: Record<string, unknown>;
  level?: string;
  description?: string;
  timing_score?: number;
  ret_1y?: number;
  ret_3y?: number;
  ret_5y?: number;
  sharpe?: number;
  ter?: number;
  volatility?: number;
  max_drawdown?: number;
  z_trend?: number;
  pullback?: number;
  momentum?: number;
}

/** POST /api/portfolio/compare-funds response */
export interface FundComparisonResult {
  funds: Array<{
    isin: string;
    name: string;
    category?: string;
    expense_ratio?: number;
    aum?: number;
    rating?: number;
    srri?: number;
    management_company?: string;
    returns?: Record<string, number>;
    metrics?: Record<string, number | null>;
    signals?: Record<string, unknown>;
    level?: string;
  }>;
  chart_data: Record<string, Array<{ date: string; price: number }>>;
}

/* ─── Portfolios types ───────────────────────────────────────────────────── */

/** Saved portfolio */
export interface SavedPortfolio {
  id: string;
  name: string;
  description?: string;
  color?: string;
  fund_count?: number;
  total_value?: number;
  funds: Array<{
    isin: string;
    name: string;
    weight: number;
  }>;
  created_at?: string;
  updated_at?: string;
}

/** Favorite fund */
export interface FavoriteFund {
  isin: string;
  name: string;
  category?: string;
  added_at?: string;
  url?: string;
}

/** POST /api/portfolio/portfolios/compare response */
export interface PortfolioComparisonResult {
  history: Record<string, Array<{ date: string; price: number }>>;
  metrics: Record<
    string,
    {
      total_return: number;
      ann_return: number;
      volatility: number;
      sharpe: number;
      max_drawdown: number;
    }
  >;
  weight_comparison?: Array<{
    fund: string;
    weight_a: number;
    weight_b: number;
  }>;
  availability?: Record<
    string,
    {
      data_start: string | null;
      fund_starts: Array<{ isin: string; name: string; first_date: string }>;
    }
  >;
}

/** Manually added position (not from CSV) */
export interface ManualPosition {
  id: number;
  isin: string;
  name: string;
  tipo: string;
  capital_invertido: number;
  valor_actual?: number | null;
  participaciones?: number | null;
  fecha_compra?: string | null;
  added_at?: string;
  updated_at?: string;
}

/** Transaction sign override (corrects CSV export omissions) */
export interface TransactionOverride {
  id: number;
  isin: string;
  fecha: string; // YYYY-MM-DD
  participaciones: number;
  notes?: string;
  created_at?: string;
  /** Enriched fields returned by the API */
  fondo?: string;                       // Resolved fund name
  actual_participaciones?: number | null; // Real value from movements (post override)
  importe?: number | null;              // Negative amount (€) from movements
}

/** Raw movement row returned by GET /api/portfolio/raw-movements */
export interface RawMovement {
  isin: string;
  fecha: string; // YYYY-MM-DD
  participaciones: number;
  importe: number;
  tipo: string;
  fondo: string;
  fuente: string;
}

/** Excluded movement returned by GET /api/portfolio/excluded-movements */
export interface ExcludedMovement {
  isin: string;
  fecha: string; // YYYY-MM-DD
  fondo: string;
  importe: number | null;
  participaciones: number | null;
}

/** Data provider status for one ISIN (GET /api/portfolio/providers-status) */
export interface ProviderStatus {
  isin: string;
  canonical: string;
  name: string;
  raw_isins: string[];
  rows: number;
  first_date: string | null;
  last_date: string | null;
  is_fresh: boolean;
  is_stale: boolean;
  no_data: boolean;
  /** Per-provider row counts: { Finect: 1234, YahooFinance: 1000, FMP: 0 } */
  providers?: Record<string, number>;
}
